src/System.Private.CoreLib/shared/System/Collections/Generic/Dictionary.cs

```diff
@@ -36,14 +39,15 @@ internal enum InsertionBehavior : byte
     [TypeForwardedFrom("mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089")]
     public class Dictionary<TKey, TValue> : IDictionary<TKey, TValue>, IDictionary, IReadOnlyDictionary<TKey, TValue>, ISerializable, IDeserializationCallback where TKey : notnull
     {
+        [StructLayout(LayoutKind.Auto)]
         private struct Entry
         {
+            public uint hashCode;
+            public TKey key;           // Key of entry
```

Comment:
Does Mono implement the auto layout algorithm to pack the structure efficiently? Heap size matters for Mono - we do not want to regress it. 

Is there a material benefit in not keeping `next` and `hashCode` together here, and type loader putting them together again in most cases?

Reply:
Seems coreclr isn't very good at packing structures efficiently with Auto?
```
Type layout for 'DictionaryAuto<short, short>.Entry'
Size: 16 bytes. Paddings: 4 bytes (%25 of empty space)
|==================================|
|   0-3: UInt32 hashCode (4 bytes) |
|----------------------------------|
|   4-7: Int32 next (4 bytes)      |
|----------------------------------|
|   8-9: Int16 key (2 bytes)       |
|----------------------------------|
| 10-11: Int16 value (2 bytes)     |
|----------------------------------|
| 12-15: padding (4 bytes)         |
|==================================|
```
vs
```
Type layout for 'DictionaryCurrent<short, short>.Entry'
Size: 12 bytes. Paddings: 0 bytes (%0 of empty space)
|==================================|
|   0-3: Int32 next (4 bytes)      |
|----------------------------------|
|   4-7: UInt32 hashCode (4 bytes) |
|----------------------------------|
|   8-9: Int16 key (2 bytes)       |
|----------------------------------|
| 10-11: Int16 value (2 bytes)     |
|==================================|
```
and
```
Type layout for 'DictionaryAuto<short, byte>.Entry'
Size: 16 bytes. Paddings: 5 bytes (%31 of empty space)
|==================================|
|   0-3: UInt32 hashCode (4 bytes) |
|----------------------------------|
|   4-7: Int32 next (4 bytes)      |
|----------------------------------|
|   8-9: Int16 key (2 bytes)       |
|----------------------------------|
|    10: Byte value (1 byte)       |
|----------------------------------|
| 11-15: padding (5 bytes)         |
|==================================|
```
vs
```
Type layout for 'DictionaryCurrent<short, byte>.Entry'
Size: 12 bytes. Paddings: 1 bytes (%8 of empty space)
|==================================|
|   0-3: Int32 next (4 bytes)      |
|----------------------------------|
|   4-7: UInt32 hashCode (4 bytes) |
|----------------------------------|
|   8-9: Int16 key (2 bytes)       |
|----------------------------------|
|    10: Byte value (1 byte)       |
|----------------------------------|
|    11: padding (1 byte)          |
|==================================|
```
Will change back but switch the order of `hashCode` and `next`

Reply:
The ideal layout would be to align each field by 4 bytes and 16 bytes in total? 
Mono should have a similar to CoreCLR behavior for Auto, will check and update here.

Reply:
> The ideal layout would be to align each field by 4 bytes and 16 bytes in total?

For `Dictionary<short, short>.Entry` that means the `Entry[]` is 25% empty space when it doesn't need that space for alignment (the top example).

The default (Sequential) makes it 12bytes which makes for nice packing; but adding Auto makes it 16bytes 😢

Reply:
array of 16 bytes structs won't cross a cache-line boundary if aligned, 12 bytes ones will

Reply:
Related: https://github.com/dotnet/coreclr/issues/25422

Reply:
> array of 16 bytes structs won't cross a cache-line boundary if aligned

That's not correct. 16 bytes structs won't cross a cache-line boundaries only if they are aligned at 16 bytes.